### DISCUSSION TITLE
Rotate, Crop, Annotate - in 1 step? or 4?

### DISCUSSION THREAD

**Discussion by GeneaCoder** [ImageMagick/ImageMagick#2077] [Help] [answered]
Version: ImageMagick 6.7.8-9 2019-06-11 Q16 http://www.imagemagick.org
Copyright: Copyright (C) 1999-2012 ImageMagick Studio LLC
Features: OpenMP 

First time poster, let me know if I'm doing it wrong.

I'm building a utility to chop down some large images to what I want. I'll give the basics as best I can. I have several hundred images to do, one at a time.

All images are going to be ~4200x3500 pixels. 99% of the images will have to be rotated ~ -5 to 5 (355 to 5 deg). All images will be cropped to ~1400x250 pixels. 95% of them will then be annotated with 4 digits of text. All that being said, I may want to expand this someday, but the basics will likely NOT change -- I rotate, then crop, then annotate.

QUESTION: Should I do this in one step? Currently I'm doing 3 steps...and considering a 4th...here's why...

PROBLEM: My server is slow. I'm on an inexpensive hosting site. It's perfect for what I do. But step 1, rotating an image that large, takes about 14-20 seconds. If that what it takes, ok. But can I make it faster by cropping BEFORE rotating, then rotate and crop again. Or is going to 1 step the answer? I don't know. I have enough knowledge about this to make me a little dangerous...that's it.

DETAILS: I'm running this via php with a Javascript front end that allows me to set the rotation, crop, and annotation. So, I can see it before sending it to the backend. 

The Javascript passes everything I need.
Rotation angle
Zoom % (if needed)
Orig size of image (in pixels)
Crop Start (x,y) on image
Crop w,h
All Annotate Info (Font, size, location, color)

My 3 steps...
Rotate (14-20 sec) - `convert orig.jpg -rotate 3.9  rotated.jpg`
(I use dimensions from rotated.jpg to compute crop, but could do this with trig if needed)
Crop (.3-1 sec) - `convert rotated.jpg -crop  1387x236+716+1298 cropped.jpg`
Annotate (.2-1.2 sec) - `convert cropped.jpg -fill '#ff0000 -pointsize 30 -font [pathto.ttf] -annotate +538+45 '1872' finished.jpg`

That's it. Thanks for reading this novel.

**Comment by snibgo** [comment, marked as answer]
Your version of IM is very very old. I suggest you upgrade it (or ask your hoster to upgrade it).

You suggest using JPEG as an intermediate format. But this is lossy, which mean every compressed write will change pixel colours. I suggest you don't use JPEG. Instead, use a lossless format. If you need intermediate files but don't need to keep them, I suggest you use MIFF or MPC. If you do need to keep them, use TIFF or PNG.

After "-crop", insert "+repage".

I expect overall performance will be improved if you crop first, before rotating.

As you say, your server is very slow. Rotating a 4200x3500 image by 3 degrees takes my (old) laptop 3.4 seconds, not 14-20 seconds.

Doing the whole job in a single command will be faster.

**Reply by GeneaCoder** [reply]
Thank you! I have and will take this advice. I went ahead and did the trig to do a precise PreCrop. It was annoying partly because it's been 35 years since I did trig and mostly because I made some dumb mistakes.  Suffice to say, it works now...and it's WAY faster. I was about to combine the last 3 (rotate, crop, & annotate) in one command with +repage, when I realized I could do all 4 (duh).

`convert original.jpg -crop [pW]x[pH]+[pX]+[pY] +repage -rotate [angle] +repage crop [W]x[H]+[X]+[Y] +repage -fill [color] -pointsize [size] -font pathtoTTF -annotate +[tX]+[tY] 'mytextstring' finished.jpg`

Not sure the repage before the annotate is necessary, but, it works -- and it's SO MUCH FASTER.

Thank you again!
-R

P.S.
I'll write this out in case anyone comes along and wants to do this.

DISCLAIMER: I will never, ever claim this is the best way. I would probably guess there's a hundred ways to do this better. This is the way I did it. If there's a better way, post it. I promise not to get mad, angry, snarky, belligerent, or ornery. I'm not that guy. Also, I still have some bugs in the system...for instance, small crops with large rotations near the center of the image get horked. Not sure why.

During development, I did this in Javascript (frontend) so I could troubleshoot it easier. The final version is done on the backend (php). I start with my X,Y starting point of the ROTATED CROP -- I rotate it so it's level and mark my crop window on the frontend. I pass this X,Y (and many more variables) to the backend. The code below is NOT exact Javascript and it's NOT exact php...but hopefully close enough to both for you to get the idea.

Some definitions...
Cartesian coordinates system - where 0,0 is in the middle and NOT in the upper left corner
Screen coordinates system - where 0,0 is in the upper left of your screen, NOT the center

First, some simple trig...
The chunk of image I have is rotated on my frontend. I want to PreCrop this. The PreCrop image will have "extra image bits" in 4 RIGHT TRIANGLES surrounding the image. Regardless of rotation, the triangles on the left and right will be the same size (just rotated 180 deg). Same with top and bottom triangles. You need to know the side lengths of these triangles. Well, we know 2 of the 3 angles, and we know the length of 1 side.

```
	pi = pi();
	AngleC = abs(CurrRotAngle);
	AngleA = 90;
	AngleB = 180 - AngleA - CurrRotAngle
	RadC = AngleC * (pi/180);
	RadA = AngleA * (pi/180);
	RadB = AngleB * (pi/180);

	LeftVert = TarCropH * sin(RadB) / sin(RadA);
	LeftHorz = TarCropH * sin(RadC) / sin(RadA);
	TopVert = TarCropW * sin(RadC) / sin(RadA);
	TopHorz = TarCropW * sin(RadB) / sin(RadA);
```

My TarCrop* is the W and H of what I eventually want. Since it's not yet "straight" the W and H are the hypotenuses. LeftVert is the vertical side of the left triangle and so forth. You'll need these later.

To do the required trig, I thought it better to switch to Cartesian. Again, not sure I needed to do this, but I did.
```
	MidPointX = img_orig_w / 2;
	MidPointY = img_orig_h / 2;
```

Now where are my coordinates?
```
	CartesianX = TarCropX - MidPointX;
	CartesianY = MidPointY - TarCropY;
```

Here's where the trig got a little heavier. The image is rotating around its center point. So, if we were to draw a circle around the center point that intersects my X and Y, then no matter how much I rotate it, that point will always be on the circle. You don't have to draw the circle, but you better believe it's there. So, how big is it? Let's get the radius...
```
	// a^2 + b^2 = c^2
	XYRadius = sqrt((abs(CartesianX) * abs(CartesianX)) + (abs(CartesianY) * abs(CartesianY)));
```
And the angle...?
```
	XYAngle = atan2(CartesianY, CartesianX) * 180 / pi; // NOTE! Y is first
	if (XYAngle < 0) {
		XYAngle = XYAngle + 360; // because javascript & php are weird
	}
```

Now, where's it going? It's going BACK to the original location.
```
	OffsetDeg = XYAngle + CurrRotAngle;
	OffsetRad = OffsetDeg * pi / 180; // converted to radians
	NewX = cos(OffsetRad) * XYRadius;
	NewY = sin(OffsetRad) * XYRadius;
```

And there you go. X and Y have been rotated around your imaginary circle and their new (technically original) locations are set. Let switch back to screen coordinates system now...
```
	NewX = NewX + MidPointX;
	NewY = MidPointY - NewY;
```

Now, here's where things went horribly wrong for me. Suffice to say that my confusion lay in the following...
Screen Coord Sys - CW rotation is generally viewed as POSITIVE -- CCW rotation is NEGATIVE
Cartesian Sys - it's just the opposite...

In any case, once I got past this snafu, all was well...
```
	if (CurrRotAngle < 0) {
		PreCropX = NewX - LeftHorz;
		PreCropY = NewY;
	} else {
		PreCropX = NewX;
		PreCropY = NewY - TopVert;
	}
	PreCropW = LeftHorz + TopHorz;
	PreCropH = LeftVert + TopVert;
```

And there you go. X,Y,W, & H for the PreCrop. On my backend, I PreCrop, Rotate, Crop & Annotate in one step.
